skbio/stats/composition.py

```diff
@@ -0,0 +1,419 @@
+       are nonzero and each composition (row) adds up to 1
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from skbio.stats.composition import multiplicative_replacement
+    >>> X = np.array([[.2,.4,.4, 0],[0,.5,.5,0]])
+    >>> multiplicative_replacement(X)
+    array([[ 0.1875,  0.375 ,  0.375 ,  0.0625],
+           [ 0.0625,  0.4375,  0.4375,  0.0625]])
+
+    """
+    mat = np.asarray(mat)
+    z_mat = (mat == 0).astype(np.float32)
+
+    if len(mat.shape) == 1:
+        num_feats = len(mat)
+        num_samps = 1
+        tot = z_mat.sum()
+    elif len(mat.shape) == 2:
+        num_samps, num_feats = mat.shape
+        tot = z_mat.sum(axis=1)
+    else:
+        raise ValueError("mat has too many dimensions")
+
+    if delta is None:
+        delta = (1. / num_feats)**2
+
+    zcnts = 1 - np.reshape(tot * delta, (num_samps, 1))
+    mat_ = _closure(z_mat*delta + np.multiply((1-z_mat),
+                                              np.multiply(zcnts, mat)))
+    if len(mat.shape) == 1:
+        mat_ = np.ravel(mat_)
+    return mat_
+
+
+def perturb(x, y):
+    r"""
+    Performs the perturbation operation
+
+    This operation is defined as
+    :math:`x \oplus y = C[x_1 y_1, ..., x_D y_D]`
+
+    :math:`C[x]` is the closure operation defined as
+    :math:`C[x] = [\frac{x_1}{\sum x},...,\frac{x_D}{\sum x}]`
+    for some :math:`D` dimensional real vector :math:`x` and
+    :math:`D` is the number of components for every composition.
+
+    Parameters
+    ----------
+    x : array-like, float
+        a matrix of proportions where
+        rows = compositions and
+        columns = components
+    y : array-like, float
+        a matrix of proportions where
+        rows = compositions and
+        columns = components
+
+    Returns
+    -------
+    mat_ : numpy.ndarray, np.float64
+       A matrix of proportions where all of the values
+       are nonzero and each composition (row) adds up to 1
+
+    Notes
+    -----
+    - Each row must add up to 1
+
+    - All of the values in x and y must be greater than zero
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from skbio.stats.composition import perturb
+    >>> x = np.array([.1,.3,.4, .2])
+    >>> y = np.array([1./6,1./6,1./3,1./3])
+    >>> perturb(x,y)
+    array([ 0.0625,  0.1875,  0.5   ,  0.25  ])
+
+    """
+    x = np.asarray(x)
+    y = np.asarray(y)
+    mat_ = _closure(np.multiply(x, y))
+    return mat_
+
+
+def perturb_inv(x, y):
+    r"""
+    Performs the inverse perturbation operation
+
+    This operation is defined as
+    :math:`x \ominus y = C[x_1 y_1^{-1}, ..., x_D y_D^{-1}]`
+
+    :math:`C[x]` is the closure operation defined as
+    :math:`C[x] = [\frac{x_1}{\sum x},...,\frac{x_D}{\sum x}]`
+    for some :math:`D` dimensional real vector :math:`x` and
+    :math:`D` is the number of components for every composition.
+
+    Parameters
+    ----------
+    x : numpy.ndarray
```

Comment:
Should be `array_like` too (also, note the underscore, not hyphen). BTW, the reference for docstrings formatting is [in numpy's repo](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#docstring-standard)